### ISSUE TITLE
3.0 - Missing filters for order item name and if possible other data in order view

### ISSUE THREAD

**Issue by napoleaowingert** [awesomemotive/easy-digital-downloads#9265]
### Expected behavior
What about adding filters to customize the order view?

eg. It seems we don't have an easy way to customize the order item name. 

### Actual behavior

Please see https://github.com/awesomemotive/easy-digital-downloads/blob/60a205cfab9136a3994b548d78f6bb2f5913628f/includes/orders/classes/class-order-item.php#L234-L236

By adding a filter there, it would be a lot easier when trying to add custom data to the order item name (eg. data from custom price option settings)

It would also be useful to add filters to other data that is sent to the order view.

If it already exists and I missed it, please let me know. Thank you.

EDD Version (or branch): 3.0

**Comment by napoleaowingert** [comment]
Hi @robincornett, thank you for checking this.

Yes, we can explain this a bit more.

We are not planning to change the database data, so we believe the entries would still be immutable.

We simply need an easy way to add more data to the order view.

For example, if you have variable priced products that contains additional data for each price option (through their advanced settings), we could add more data to the order item name.

The reason is that sometimes only the price option name might not properly define a unique price option when having extra data for them. So having extra data as part of the order item name would do the trick.

So when checking the order details, the store owner quickly understands all order items directly in that view.

Without having this, it creates a lot of extra manual work when checking the order details, as we would need to open the product page of each order item just to check the additional data.

By using that filter to change the product name before it's added to the database as part of an order, we would need to combine it with another hook, to only apply that change in that single order details view. Is that correct?

Which would be the additional hook?

Maybe it would be easier and better to use a filter for the order item name directly?

**Comment by cklosowski** [comment]
@napoleaowingert Is there a filter you are using already to achieve your goal that we can look at and maybe see how it can be implemented in EDD 3.0 or is this a wishlist item?

I'm more asking to know if we've introduced a 'regression' and if so, if it's an acceptable regression that has an alternative or if we have something that is just no longer possible (that we need to look into).

Thanks for your detailed response.

**Comment by napoleaowingert** [comment]
Hi @cklosowski, thank you for checking this.

Yes, we are using Easy Digital Downloads version 2.11.7 and it works.

That version does not have a direct filter as well, but we can achieve this goal by combining hooks.

The order item name is built by joining the download name with the price option name, see:

https://github.com/awesomemotive/easy-digital-downloads/blob/master/includes/admin/payments/view-order-details.php#L459-L468

It builds the order item name by using the download name with the price option name.

So we can use the **edd_get_price_option_name** hook (filter) to add more data to the final order item name.

In order to avoid using that filter everywhere, we restrict that by using other hooks - one to add the filter and another to remove it.

Please see the example below:

```

function myprefix_view_order_details_filter_edd_get_price_option_name( $price_name, $download_id, $payment_id, $price_id ) {
	// Set the price name
	$price_name = "{$price_name} - Hello World";

	return $price_name;
}

function myprefix_view_order_details_action_edd_view_order_details_main_before( $payment_id ) {
	add_filter( 'edd_get_price_option_name', 'myprefix_view_order_details_filter_edd_get_price_option_name', 10, 4 );
}
add_action( 'edd_view_order_details_main_before', 'myprefix_view_order_details_action_edd_view_order_details_main_before' );

function myprefix_view_order_details_action_edd_view_order_details_files_after( $payment_id ) {
	remove_filter( 'edd_get_price_option_name', 'myprefix_view_order_details_filter_edd_get_price_option_name' );
}
add_action( 'edd_view_order_details_files_after', 'myprefix_view_order_details_action_edd_view_order_details_files_after' );

```

**Comment by cklosowski** [comment]
Thanks for the details @napoleaowingert. So one challenge here is that a filter won't do much good on the row since it is rendered with Backbone JS, so a PHP filter won't even have the context available yet, 
```
			<a
				href="<?php echo esc_url( $view_url ); ?>&post={{ data.productId }}"
				class="row-title"
			>
				{{{ data.productName }}}
			</a>
```

Let me see what I can come up with for an option here.

**Comment by cklosowski** [comment]
I think I found a way to add a filter JUST to the admin view order details rendering, which I'm more comfortable with than allowing filtering the data directly within the `Order_Item` class itself. We'll test this out a bit, and see if we can find anything I haven't thought of yet.

**Comment by cklosowski** [comment]
Tagging as a 'regression' b/c this was a place that was filterable before, but was not now. Long term, I believe we will need to find a way to have 'additional' data output as a supported concept as I'd like to use it for Recurring and Software Licensing.

**Comment by cklosowski** [comment]
@napoleaowingert Alright! here is your filter:
`apply_filters( 'edd_order_details_item_name', $this->product_name, $this );`

This filter will ONLY run in the admin area, so need to dance around hooks and filters to restrict it to the admin.

**Comment by WildeWebMarketing** [comment]
Yes, a filter would definitely be useful or, in my case, a way to turn off the option name being added to the base name now. For example, when using the [downloads_history] shortcode, my customers only need to see the base name. So now I've had to edit each template file and use str_replace to remove it.
